test/ClangImporter/availability_app_extension.swift

```diff
@@ -5,5 +5,5 @@
 import Foundation
 
 func test_unavailable_app_extension() {
-  _ = SomeCrazyAppExtensionForbiddenAPI() // expected-error {{'SomeCrazyAppExtensionForbiddenAPI()' is unavailable: Not available in App Extensions}}
+  _ = SomeCrazyAppExtensionForbiddenAPI() // expected-error {{'SomeCrazyAppExtensionForbiddenAPI()' is unavailable on OS X application extension: Not available in App Extensions}}
```

Comment:
Hm, that is…not great. "in an OS X application extension" / "in a macOS application extension" / "in macOS application extensions" would be better.

Reply:
I don't see a good way to tweak just this diagnostic, so let's talk about a more systematic change. As this PR stands, we have:

```
// diag::availability_decl_only_version_newer
%0 is only available on macOS 10.10 or newer
%0 is only available on macOS application extension 10.10 or newer

// diag::availability_accessor_only_version_newer
setter for %1 is only available on macOS 10.10
setter for %1 is only available on macOS application extension 10.10

// diag::availability_inout_accessor_only_version_newer
cannot pass as inout because setter for %1 is only available on macOS 10.10 or newer
cannot pass as inout because setter for %1 is only available on macOS application extension 10.10 or newer

// diag::availability_deprecated, unconditional
%1 is deprecated on macOS
%1 is deprecated on macOS application extension

// diag::availability_deprecated, conditional on version
%1 was deprecated in macOS 10.10
%1 was deprecated in macOS application extension 10.10

// diag::availability_decl_unavailable
%1 is unavailable on macOS
%1 is unavailable on macOS application extension

// diag::availability_obsoleted
%1 was obsoleted in macOS 10.10
%1 was obsoleted in macOS application extension 10.10

// no, not diag::availability_introduced_in_version, that's only used for Swift
// and PackageDescription

// diag::availability_protocol_requires_version
protocol %0 requires %1 to be available on macOS 10.10 and newer
protocol %0 requires %1 to be available on macOS application extension 10.10 and newer

// diag::serialization_target_too_new
compiling for macOS 10.10, but module %2 has a minimum deployment target of macOS 10.12: %4
compiling for macOS application extension 10.10, but module %2 has a minimum deployment target of macOS application extension 10.12: %4
```

(I'm assuming that any message with a platform string in it could have an app extension platform; that's probably not true, particularly for the last one.)

Suppose we change the platform name to "application extension for macOS" and standardize on "in" instead of "on" when referring to a platform. That gives us:

```
// diag::availability_decl_only_version_newer (*)
%0 is only available in macOS 10.10 or newer
%0 is only available in application extension for macOS 10.10 or newer

// diag::availability_accessor_only_version_newer (*)
setter for %1 is only available in macOS 10.10
setter for %1 is only available in application extension for macOS 10.10

// diag::availability_inout_accessor_only_version_newer (*)
cannot pass as inout because setter for %1 is only available in macOS 10.10 or newer
cannot pass as inout because setter for %1 is only available in application extension for macOS 10.10 or newer

// diag::availability_deprecated, unconditional (*)
%1 is deprecated in macOS
%1 is deprecated in application extension for macOS

// diag::availability_deprecated, conditional on version
%1 was deprecated in macOS 10.10
%1 was deprecated in application extension for macOS 10.10

// diag::availability_decl_unavailable (*)
%1 is unavailable in macOS
%1 is unavailable in application extension for macOS

// diag::availability_obsoleted
%1 was obsoleted in macOS 10.10
%1 was obsoleted in application extension for macOS 10.10

// no, not diag::availability_introduced_in_version, that's only used for Swift
// and PackageDescription

// diag::availability_protocol_requires_version (*)
protocol %0 requires %1 to be available in macOS 10.10 and newer
protocol %0 requires %1 to be available in application extension for macOS 10.10 and newer

// diag::serialization_target_too_new
compiling for macOS 10.10, but module %2 has a minimum deployment target of macOS 10.12: %4
compiling for application extension for macOS 10.10, but module %2 has a minimum deployment target of application extension for macOS 10.12: %4
```

Do you think those all read well?

Reply:
Application extensions have always gotten less attention, but judging from how few diagnostics you've had to change, it might not have mattered much in the past. What if we use "application extension**s** for macOS"? That's a weird name for a platform, but it'll work well enough here.

If not, it might be worth adding an `isApplicationExtensionVariant` for platforms (if there isn't one already), and treating that differently somehow.